core/PRU.cpp

```diff
@@ -770,6 +778,10 @@ int PRU::start(char * const filename, const McaspRegisters& mcaspRegisters)
 			return 1;
 		}
 	}
+#endif
+#if ENABLE_PRU_RPROC == 1
+// do something else probably? Or simply not required.
```

Comment:
This is where we should be calling  `pruManager->start()`.

I guess the issue here is how to handle `prussdrv`'s flexibility of executing embedded vs external binary code. I suggest the following:
- add a `std::string` argument:  `PruManager::start(const std::string& path)`.
- if (`path == ""`), `Uio` will call `prussdrv_exec_code()`, while `Rproc` will use the default path of the binary built by the `pasm/clpru` workflow.
- if a non-empty string is passed, `Uio` will call `prussdrv_exec_program()`, while `Rproc` will use the provided path, which is assumed to be that of a file that was built with the `pasm/clpru` workflow.

One issue remains here, that of what happens for `pru_rtaudio.p` vs  `pru_rtaudio_irq.p`, but given how we are hoping to phase out the former soon, we could just hardcode either for now.

The code here would then look something like:

```
	if(filename[0] == '\0') { //if the string is empty, load the embedded code
		if(gRTAudioVerbose)
			printf("Using embedded PRU code\n");
		if(pruManager->start()) {
			fprintf(stderr, "Failed to execute PRU code\n");
			return 1;
		}
	} else {
		if(gRTAudioVerbose)
			printf("Using PRU code from %s\n",filename);
		if(pruManager->start(filename)) {
			fprintf(stderr, "Failed to execute PRU code from %s\n", filename);
			return 1;
		}
	}
```
which could be further simplified to:
	
```
	if(gRTAudioVerbose)
		printf("Using %s PRU code\n", 0 == strlen(filename) ? "embedded" : filename);
	if(pruManager->start(filename)) {
		fprintf(stderr, "Failed to execute PRU code\n");
		return 1;
	}
```

Reply:
Pardon the tone, but the comments in some places [like this one](https://github.com/giuliomoro/Bela-dhruva/blob/multi-codec/core/PRU.cpp#L741) are a bit non- self-explanatory, What exactly does _avoid stupid compiler warning_  mean? Also, if we are type casting, then I suggest we use `static_caste<int>` instead as I have learned it's a better practice from here: https://www.youtube.com/watch?v=HlNVgmvX1EI&t=451s 
Also, `prussdrv_exec_code(pru_number, pruCode, pruCodeSize))` has 3 input parameters unlike `prussdrv_exec_program(pru_number, filename))` which has only 2. So should I do function overloading by defining 2 different `start()` functions then? (To be noted, the overloading will prove to be useless in rproc and one of the overloaded function will have a blank definition.)
Or, as [you said earlier in this comment,](https://github.com/giuliomoro/Bela-dhruva/pull/1#discussion_r672162925) we could just put the  `prussdrv_exec_code()` part in the class implementation and leave [exec_program()](https://github.com/giuliomoro/Bela-dhruva/blob/multi-codec/core/PRU.cpp#L766-L770) as is. Again, this wouldn't be ideal because on non-uniformity in the code.

Reply:
>  as you said earlier in this comment, we could just put the prussdrv_exec_code() part in the class implementation and leave exec_program() as is.

that's not quite what I meant.

> Since we can just hardcode either for now, then I can include "PruBinary.h" in PruManager.cpp and this way we can have the values of pruCode, pruCodeSize right in the start function itself

that's more like what I meant.

Reply:
> that's more like what I meant

Okay, please check the latest commit: https://github.com/giuliomoro/Bela-dhruva/pull/1/commits/a450efa930d325e0172b2c39ef2bf7aa6442618f